File: fast-and-slow-pointers/palindrome-linked-list.cpp
#include <bits/stdc++.h>

using namespace std;

struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
  public:
    bool isPalindrome(ListNode *head) {
      if (!head || !head->next)
        return true;

      ListNode *fast = head, *slow = head;

      while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
      }

      ListNode *reversed = reverseList(slow);

      while (reversed != nullptr) {
        if (head->val != reversed->val)
          return false;

        head = head->next;
        reversed = reversed->next;
      }

      return true;
    }

  private:
    ListNode* reverseList(ListNode* head) {
      ListNode *current = head, *previous = nullptr, *follow;

      while (current != nullptr) {
        follow = current->next;
        current->next = previous;
        previous = current;
        current = follow;
      }

      return previous;
    }
};

File: fast-and-slow-pointers/palindrome_linked_list.py
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        def getMiddle(head: ListNode):
            fast = slow = head

            while fast and fast.next:
                slow = slow.next
                fast = fast.next.next

            return slow

        def reverse(head: ListNode):
            prev, curr = None, head

            while curr:
                next = curr.next
                curr.next = prev
                prev = curr
                curr = next

            return prev

        # Reverse the second half
        middle = reverse(getMiddle(head))

        while middle:
            if head.val != middle.val:
                return False
            else:
                head = head.next
                middle = middle.next

        return True